File: ExamMinire.cpp
#include <iostream>
#include <vector>
#include <string>
#include <iomanip>
#include <unordered_map>
#include <stack>
#include <queue>

using namespace std;

struct Estudiante
{
    string codigo;
    string nombre;
    float nota1;
    float nota2;
    float nota3;
    float promedio;
};

vector<Estudiante> hoja;
unordered_map<string, int> indiceHash;
stack<vector<Estudiante>> historial;
queue<string> colaAtencion;

float calcularPromedio(float n1, float n2, float n3)
{
    return (n1 + n2 + n3) / 3.0f;
}

void actualizarHash()
{
    indiceHash.clear();

    for (int i = 0; i < (int)hoja.size(); i++)
    {
        indiceHash[hoja[i].codigo] = i;
    }
}

void guardarHistorial()
{
    historial.push(hoja);
}

void registrarEstudiante()
{
    Estudiante e;

    cout << "\nCodigo: ";
    cin >> e.codigo;

    if (indiceHash.count(e.codigo))
    {
        cout << "Codigo ya registrado.\n";
        return;
    }

    cin.ignore();

    cout << "Nombre: ";
    getline(cin, e.nombre);

    cout << "Nota 1: ";
    cin >> e.nota1;

    cout << "Nota 2: ";
    cin >> e.nota2;

    cout << "Nota 3: ";
    cin >> e.nota3;

    e.promedio = calcularPromedio(e.nota1, e.nota2, e.nota3);

    guardarHistorial();

    hoja.push_back(e);

    actualizarHash();

    cout << "Estudiante registrado.\n";
}

void mostrarHoja()
{
    if (hoja.empty())
    {
        cout << "\nNo hay registros.\n";
        return;
    }

    cout << "\n";
    cout << left
         << setw(12) << "Codigo"
         << setw(25) << "Nombre"
         << setw(8) << "N1"
         << setw(8) << "N2"
         << setw(8) << "N3"
         << setw(10) << "Promedio"
         << endl;

    for (const auto &e : hoja)
    {
        cout << left
             << setw(12) << e.codigo
             << setw(25) << e.nombre
             << setw(8) << e.nota1
             << setw(8) << e.nota2
             << setw(8) << e.nota3
             << setw(10) << fixed << setprecision(2)
             << e.promedio
             << endl;
    }
}

void buscarEstudiante()
{
    string codigo;

    cout << "\nCodigo a buscar: ";
    cin >> codigo;

    auto it = indiceHash.find(codigo);

    if (it == indiceHash.end())
    {
        cout << "No encontrado.\n";
        return;
    }

    Estudiante &e = hoja[it->second];

    cout << "\nNombre: " << e.nombre;
    cout << "\nPromedio: " << e.promedio << endl;
}

int particion(vector<Estudiante> &arr, int bajo, int alto)
{
    float pivote = arr[alto].promedio;

    int i = bajo - 1;

    for (int j = bajo; j < alto; j++)
    {
        if (arr[j].promedio >= pivote)
        {
            i++;
            swap(arr[i], arr[j]);
        }
    }

    swap(arr[i + 1], arr[alto]);

    return i + 1;
}

void quickSort(vector<Estudiante> &arr, int bajo, int alto)
{
    if (bajo < alto)
    {
        int pi = particion(arr, bajo, alto);

        quickSort(arr, bajo, pi - 1);
        quickSort(arr, pi + 1, alto);
    }
}

void merge(vector<Estudiante> &arr, int izq, int medio, int der)
{
    int n1 = medio - izq + 1;
    int n2 = der - medio;

    vector<Estudiante> L(n1);
    vector<Estudiante> R(n2);

    for (int i = 0; i < n1; i++)
        L[i] = arr[izq + i];

    for (int j = 0; j < n2; j++)
        R[j] = arr[medio + 1 + j];

    int i = 0;
    int j = 0;
    int k = izq;

    while (i < n1 && j < n2)
    {
        if (L[i].codigo <= R[j].codigo)
            arr[k++] = L[i++];
        else
            arr[k++] = R[j++];
    }

    while (i < n1)
        arr[k++] = L[i++];

    while (j < n2)
        arr[k++] = R[j++];
}

void mergeSort(vector<Estudiante> &arr, int izq, int der)
{
    if (izq < der)
    {
        int medio = (izq + der) / 2;

        mergeSort(arr, izq, medio);
        mergeSort(arr, medio + 1, der);

        merge(arr, izq, medio, der);
    }
}

void ordenarPorPromedio()
{
    if (hoja.empty())
        return;

    guardarHistorial();

    quickSort(hoja, 0, hoja.size() - 1);

    actualizarHash();

    cout << "Ordenado por promedio.\n";
}

void ordenarPorCodigo()
{
    if (hoja.empty())
        return;

    guardarHistorial();

    mergeSort(hoja, 0, hoja.size() - 1);

    actualizarHash();

    cout << "Ordenado por codigo.\n";
}

void deshacer()
{
    if (historial.empty())
    {
        cout << "Nada que deshacer.\n";
        return;
    }

    hoja = historial.top();
    historial.pop();

    actualizarHash();

    cout << "Operacion deshecha.\n";
}

void agregarCola()
{
    string codigo;

    cout << "\nCodigo: ";
    cin >> codigo;

    if (!indiceHash.count(codigo))
    {
        cout << "No existe.\n";
        return;
    }

    colaAtencion.push(codigo);

    cout << "Agregado a la cola.\n";
}

void atenderCola()
{
    if (colaAtencion.empty())
    {
        cout << "Cola vacia.\n";
        return;
    }

    string codigo = colaAtencion.front();
    colaAtencion.pop();

    cout << "Atendiendo: "
         << hoja[indiceHash[codigo]].nombre
         << endl;
}

float sumaRecursiva(int i)
{
    if (i >= (int)hoja.size())
        return 0;

    return hoja[i].promedio + sumaRecursiva(i + 1);
}

void estadisticas()
{
    if (hoja.empty())
    {
        cout << "Sin datos.\n";
        return;
    }

    float suma = sumaRecursiva(0);

    cout << "\nCantidad: "
         << hoja.size();

    cout << "\nPromedio General: "
         << fixed << setprecision(2)
         << suma / hoja.size()
         << endl;
}

void menu()
{
    int opcion;

    do
    {
        cout << "\n==============================\n";
        cout << "1. Registrar estudiante\n";
        cout << "2. Mostrar hoja\n";
        cout << "3. Buscar estudiante\n";
        cout << "4. Ordenar por promedio\n";
        cout << "5. Ordenar por codigo\n";
        cout << "6. Deshacer\n";
        cout << "7. Agregar a cola\n";
        cout << "8. Atender cola\n";
        cout << "9. Estadisticas\n";
        cout << "10. Salir\n";
        cout << "Opcion: ";

        cin >> opcion;

        switch (opcion)
        {
        case 1: registrarEstudiante(); break;
        case 2: mostrarHoja(); break;
        case 3: buscarEstudiante(); break;
        case 4: ordenarPorPromedio(); break;
        case 5: ordenarPorCodigo(); break;
        case 6: deshacer(); break;
        case 7: agregarCola(); break;
        case 8: atenderCola(); break;
        case 9: estadisticas(); break;
        case 10: cout << "Fin del programa.\n"; break;
        default: cout << "Opcion invalida.\n";
        }

    } while (opcion != 10);
}

int main()
{
    menu();
    return 0;
}

File: EXAMminire.py
from collections import deque
from copy import deepcopy

# ==========================
# ESTRUCTURAS DE DATOS
# ==========================

hoja = []
indice_hash = {}
historial = []
cola_atencion = deque()


# ==========================
# FUNCIONES AUXILIARES
# ==========================

def calcular_promedio(n1, n2, n3):
    return (n1 + n2 + n3) / 3.0


def actualizar_hash():
    indice_hash.clear()

    for idx, estudiante in enumerate(hoja):
        indice_hash[estudiante["codigo"]] = idx


def guardar_historial():
    historial.append(deepcopy(hoja))


# ==========================
# CRUD ESTUDIANTES
# ==========================

def registrar_estudiante():
    print("\n--- REGISTRAR NUEVO ESTUDIANTE ---")

    codigo = input("Codigo unico: ").strip()

    if codigo in indice_hash:
        print("[ERROR] El codigo ya existe.")
        return

    nombre = input("Apellidos y Nombres: ").strip()

    try:
        nota1 = float(input("Nota 1 (0-20): "))
        nota2 = float(input("Nota 2 (0-20): "))
        nota3 = float(input("Nota 3 (0-20): "))
    except ValueError:
        print("[ERROR] Debe ingresar numeros validos.")
        return

    estudiante = {
        "codigo": codigo,
        "nombre": nombre,
        "nota1": nota1,
        "nota2": nota2,
        "nota3": nota3,
        "promedio": round(calcular_promedio(nota1, nota2, nota3), 2)
    }

    guardar_historial()

    hoja.append(estudiante)

    actualizar_hash()

    print("[EXITO] Estudiante registrado correctamente.")


def mostrar_hoja():

    if not hoja:
        print("\n[AVISO] No existen registros.")
        return

    print("\n" + "=" * 85)
    print(f"{'MINI HOJA DE CALCULO ACADEMICA':^85}")
    print("=" * 85)

    print(
        f"{'Codigo':<12}"
        f"{'Nombre':<30}"
        f"{'N1':<10}"
        f"{'N2':<10}"
        f"{'N3':<10}"
        f"{'Promedio':<10}"
    )

    print("-" * 85)

    for e in hoja:
        print(
            f"{e['codigo']:<12}"
            f"{e['nombre']:<30}"
            f"{e['nota1']:<10.1f}"
            f"{e['nota2']:<10.1f}"
            f"{e['nota3']:<10.1f}"
            f"{e['promedio']:<10.2f}"
        )

    print("=" * 85)


def buscar_estudiante():

    codigo = input(
        "\nIngrese codigo a buscar: "
    ).strip()

    if codigo in indice_hash:

        pos = indice_hash[codigo]

        e = hoja[pos]

        print("\n[ENCONTRADO]")
        print("Nombre   :", e["nombre"])
        print("Promedio :", e["promedio"])

    else:
        print("[AVISO] Codigo no encontrado.")


# ==========================
# QUICKSORT (PROMEDIO)
# ==========================

def quicksort(lista):

    if len(lista) <= 1:
        return lista

    pivote = lista[-1]

    mayores = [
        x for x in lista[:-1]
        if x["promedio"] >= pivote["promedio"]
    ]

    menores = [
        x for x in lista[:-1]
        if x["promedio"] < pivote["promedio"]
    ]

    return quicksort(mayores) + [pivote] + quicksort(menores)


# ==========================
# MERGESORT (CODIGO)
# ==========================

def merge_sort(lista):

    if len(lista) <= 1:
        return lista

    medio = len(lista) // 2

    izquierda = merge_sort(lista[:medio])
    derecha = merge_sort(lista[medio:])

    return merge(izquierda, derecha)


def merge(izquierda, derecha):

    resultado = []

    i = 0
    j = 0

    while i < len(izquierda) and j < len(derecha):

        if izquierda[i]["codigo"] <= derecha[j]["codigo"]:
            resultado.append(izquierda[i])
            i += 1
        else:
            resultado.append(derecha[j])
            j += 1

    resultado.extend(izquierda[i:])
    resultado.extend(derecha[j:])

    return resultado


# ==========================
# ORDENAMIENTO
# ==========================

def ordenar_por_promedio():

    global hoja

    if not hoja:
        print("[INFO] No hay datos.")
        return

    guardar_historial()

    hoja = quicksort(hoja)

    actualizar_hash()

    print(
        "[OK] Ordenado por promedio "
        "(QuickSort Descendente)."
    )


def ordenar_por_codigo():

    global hoja

    if not hoja:
        print("[INFO] No hay datos.")
        return

    guardar_historial()

    hoja = merge_sort(hoja)

    actualizar_hash()

    print(
        "[OK] Ordenado por codigo "
        "(MergeSort Ascendente)."
    )


# ==========================
# PILA (STACK)
# ==========================

def deshacer():

    global hoja

    if historial:

        hoja = historial.pop()

        actualizar_hash()

        print(
            "[REVERSION] "
            "Ultimo cambio restaurado."
        )

    else:
        print("[INFO] No hay historial.")


# ==========================
# COLA FIFO
# ==========================

def agregar_cola_atencion():

    codigo = input(
        "Codigo del estudiante: "
    ).strip()

    if codigo not in indice_hash:
        print("[ERROR] Codigo inexistente.")
        return

    cola_atencion.append(codigo)

    print(
        "[COLA] Estudiante agregado "
        "a la cola de espera."
    )


def atender_estudiante():

    if not cola_atencion:
        print("[INFO] Cola vacia.")
        return

    codigo = cola_atencion.popleft()

    pos = indice_hash[codigo]

    print(
        f"[ATENCION] "
        f"{hoja[pos]['nombre']}"
    )


# ==========================
# RECURSIVIDAD
# ==========================

def suma_promedios_recursiva(i):

    if i == len(hoja):
        return 0

    return (
        hoja[i]["promedio"]
        + suma_promedios_recursiva(i + 1)
    )


def estadisticas():

    if not hoja:
        print("[AVISO] No existen datos.")
        return

    suma = suma_promedios_recursiva(0)

    promedio_general = suma / len(hoja)

    mejor = max(
        hoja,
        key=lambda x: x["promedio"]
    )

    peor = min(
        hoja,
        key=lambda x: x["promedio"]
    )

    print("\n" + "=" * 40)

    print(
        f"Estudiantes registrados : "
        f"{len(hoja)}"
    )

    print(
        f"Promedio general        : "
        f"{promedio_general:.2f}"
    )

    print(
        f"Mejor estudiante        : "
        f"{mejor['nombre']} "
        f"({mejor['promedio']:.2f})"
    )

    print(
        f"Menor promedio          : "
        f"{peor['nombre']} "
        f"({peor['promedio']:.2f})"
    )

    print("=" * 40)


# ==========================
# MENU
# ==========================

def menu():

    while True:

        print("\n" + "=" * 45)
        print("MENU PRINCIPAL")
        print("=" * 45)

        print("1. Registrar estudiante")
        print("2. Mostrar hoja")
        print("3. Buscar estudiante")
        print("4. Ordenar por promedio")
        print("5. Ordenar por codigo")
        print("6. Deshacer")
        print("7. Agregar a cola")
        print("8. Atender estudiante")
        print("9. Estadisticas")
        print("10. Salir")

        print("-" * 45)

        opcion = input(
            "Seleccione opcion: "
        ).strip()

        if opcion == "1":
            registrar_estudiante()

        elif opcion == "2":
            mostrar_hoja()

        elif opcion == "3":
            buscar_estudiante()

        elif opcion == "4":
            ordenar_por_promedio()

        elif opcion == "5":
            ordenar_por_codigo()

        elif opcion == "6":
            deshacer()

        elif opcion == "7":
            agregar_cola_atencion()

        elif opcion == "8":
            atender_estudiante()

        elif opcion == "9":
            estadisticas()

        elif opcion == "10":
            print("\nFinalizando programa...")
            break

        else:
            print("[ERROR] Opcion invalida.")


# ==========================
# INICIO
# ==========================

if __name__ == "__main__":
    menu()
